core/tests/puppeteer-acceptance-tests/functions/is-element-clickable.ts

```diff
@@ -0,0 +1,207 @@
+    if (overlappingElement) {
+      overlappingElements.push(overlappingElement);
+    }
+    if (overlappingRect) {
+      overlappingElements.push(overlappingRect);
+    }
+    return overlappingElements;
+  };
+
+  /**
+   * This function is recursive and checks if the target element is the only
+   * overlapping element meaning there are no other elements blocking
+   * the target element.
+   *
+   * @param {Element} element The target element to compare with the other
+   * overlapping elements.
+   * @param {Element[]} overlappingElements The overlapping elements to compare
+   * against the target element.
+   * @returns {boolean} Whether the target element is the only overlapping
+   * element.
+   */
+  const isElementNotOverlappedByOtherElements = (
+    element: Element,
+    overlappingElements: Element[]
+  ): boolean => {
+    if (
+      overlappingElements.some(el => el === element || element.contains(el))
+    ) {
+      return true;
+    }
+
+    let elementsWithShadow = [...new Set(overlappingElements)];
+    elementsWithShadow = elementsWithShadow.filter(
+      el => el && el.shadowRoot && el.shadowRoot.elementFromPoint
+    );
+
+    let shadowElements: Element[] = [];
+    for (const shadowEl of elementsWithShadow) {
+      if (shadowEl.shadowRoot) {
+        shadowElements.push(
+          ...getTopmostOverlappingElements(element, shadowEl.shadowRoot)
+        );
+      }
+    }
+    shadowElements = [...new Set(shadowElements)];
+    shadowElements = shadowElements.filter(
+      el => !overlappingElements.includes(el)
+    );
+
+    if (shadowElements.length === 0) {
+      return false;
+    }
+
+    return isElementNotOverlappedByOtherElements(element, shadowElements);
+  };
+
+  /**
+   * This function checks if the given element is in the viewport.
+   *
+   * @param {Element} element The element to check if it is in the viewport.
+   * @returns {boolean} Whether the element is in the viewport.
+   */
+  const isElementInViewport = (element: Element): boolean => {
+    const elementDimensions = element.getBoundingClientRect();
+
+    const windowHeight =
+      window.innerHeight || document.documentElement.clientHeight;
+    const windowWidth =
+      window.innerWidth || document.documentElement.clientWidth;
+
+    const verticalInView =
+      elementDimensions.top <= windowHeight &&
+      elementDimensions.top + elementDimensions.height > 0;
+    const horizontalInView =
+      elementDimensions.left <= windowWidth &&
+      elementDimensions.left + elementDimensions.width > 0;
+
+    return verticalInView && horizontalInView;
+  };
+
+  /**
+   * This function checks if an element is a disableable element and if it is
+   * then it checks if it is disabled.
+   */
+  const isElementDisabled = (element: Element): boolean => {
+    if (
+      element instanceof HTMLFormElement ||
+      element instanceof HTMLButtonElement
+    ) {
+      return element.disabled;
+    }
+    return false;
```

Comment:
Nit: Consider refactoring to `return (element instanceof HTMLFormElement || element instanceof HTMLButtonElement) && element.disabled;`

Reply:
Done